### PR TITLE
dnn: fix HAVE_TIMVX macro definition in dnn test

### PR DESCRIPTION
**Pull request by fengyuentau**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/dnn/CMakeLists.txt [hunk 1]
```diff
@@ -295,6 +295,13 @@ if(TARGET ocv.3rdparty.cann AND OPENCV_TEST_DNN_CANN)
   endif()
 endif()
 
+ocv_option(OPENCV_TEST_DNN_TIMVX "Build test with TIM-VX" (TARGET ocv.3rdparty.timvx))
+if(TARGET ocv.3rdparty.timvx AND OPENCV_TEST_DNN_TIMVX)
+  if(TARGET opencv_test_dnn)
+    ocv_target_link_libraries(opencv_test_dnn ocv.3rdparty.timvx)
```
Comment (asmorkalov) [inline modules/dnn/CMakeLists.txt:301]: Why do you need it? It's internal detail of implementation and user code should not depend on it, including tests. libopencv_dnn should have dependency from TIMVX.

Comment (fengyuentau) [reply modules/dnn/CMakeLists.txt:301]: Test does not work with missing `HAVE_TIMVX` macro:

https://github.com/opencv/opencv/blob/240b245105bdb69b2b624ce24f08373c2b16bfde/modules/dnn/test/test_int8_layers.cpp#L11-L22

Basically I need `HAVE_TIMVX=1`.

Comment (opencv-alalek) [reply modules/dnn/CMakeLists.txt:301]: We don't need that.

There is registry API:

https://github.com/opencv/opencv/blob/240b245105bdb69b2b624ce24f08373c2b16bfde/modules/dnn/include/opencv2/dnn/dnn.hpp#L125-L126

And usage in tests:

https://github.com/opencv/opencv/blob/240b245105bdb69b2b624ce24f08373c2b16bfde/modules/dnn/test/test_common.impl.hpp#L291-L296

Comment (fengyuentau) [reply modules/dnn/CMakeLists.txt:301]: `DNN_BACKEND_TIMVX` is not in opencv/modules/dnn/test/test_common.impl.hpp.

**Comment by asmorkalov** [comment]
The mentioned macro is used in tests in 2 places and both of them may be replaced with runtime check like Alexander suggested.

**Comment by fengyuentau** [comment]
So why these happen? See

https://github.com/opencv/opencv/blob/6e4280ea81b59c6dca45bb9801b758377beead55/modules/dnn/CMakeLists.txt#L290-L295

and

https://github.com/opencv/opencv/blob/6e4280ea81b59c6dca45bb9801b758377beead55/modules/dnn/test/test_int8_layers.cpp#L11-L22

(see line 18).

---

The TIM-VX backend is only for int8-quantized models. `dnnBackendsAndTargetsInt8` is only defined and used in `modules/dnn/test/test_int8_layers.cpp`.

**Comment by opencv-alalek** [comment]
> So why these happen? See

Because tests use OpenVINO API directly to test IR models (as a reference result).

https://github.com/opencv/opencv/blob/6e4280ea81b59c6dca45bb9801b758377beead55/modules/dnn/test/test_ie_models.cpp#L185-L210

Tests doesn't use special TIM-VX API

File: modules/dnn/CMakeLists.txt [hunk 2]
```diff
@@ -295,6 +295,13 @@ if(TARGET ocv.3rdparty.cann AND OPENCV_TEST_DNN_CANN)
   endif()
 endif()
 
+ocv_option(OPENCV_TEST_DNN_TIMVX "Build test with TIM-VX" (HAVE_TIMVX))
+if(OPENCV_TEST_DNN_TIMVX)
+  if(TARGET opencv_test_dnn)
+  ocv_target_compile_definitions(opencv_test_dnn PRIVATE "HAVE_TIMVX=1")
```
Comment (opencv-alalek) [inline modules/dnn/CMakeLists.txt:301]: indentation

**Review by opencv-alalek — APPROVED** [review]

File: modules/dnn/CMakeLists.txt (hunk 2 above)
Comment (fengyuentau) [reply modules/dnn/CMakeLists.txt:301]: Thank you for finding this! Meanwhile please hold merging this PR until the hardware is in the CI running the tests.

**Comment by fengyuentau** [comment]
The log shows tim-vx backend works fine with this patch, https://github.com/opencv/opencv/actions/runs/6586141510/job/17893901244?pr=24425

Ready to be merged once all tests are done.

File: modules/dnn/CMakeLists.txt (hunk 2 above)
Comment (fengyuentau) [reply modules/dnn/CMakeLists.txt]: Resolved.

**Comment by asmorkalov** [comment]
@fengyuentau I see VIM3 build passed on CI. Is it ready? May I merge the pr?

**Review by asmorkalov — APPROVED** [review]
:+1:

**Comment by fengyuentau** [comment]
> @fengyuentau I see VIM3 build passed on CI. Is it ready? May I merge the pr?

Yes, it is ready and you can merge it.

**Pull request merged**